javalib/src/main/scala/java/lang/StrictMath.scala

```diff
@@ -0,0 +1,402 @@
+/*
+ * Scala.js (https://www.scala-js.org/)
+ *
+ * Copyright EPFL.
+ *
+ * Licensed under Apache License 2.0
+ * (https://www.apache.org/licenses/LICENSE-2.0).
+ *
+ * See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.
+ */
+
+/*
+ * ====================================================
+ * Copyright (C) 1993 by Sun Microsystems, Inc. All rights reserved.
+ *
+ * Developed at SunSoft, a Sun Microsystems, Inc. business.
+ * Permission to use, copy, modify, and distribute this
+ * software is freely granted, provided that this notice
+ * is preserved.
+ * ====================================================
+ */
+
+package java.lang
+
+import java.lang.{Double => JDouble}
+
+object StrictMath {
+
+  // Port from https://www.netlib.org/fdlibm/e_log.c
+  def log(x: scala.Double): scala.Double = {
+    val Ln2High = 6.93147180369123816490e-01
+    val Ln2Low = 1.90821492927058770002e-10
+    val Two54 = 1.80143985094819840000e+16 // 2^54
+    val Lg1 = 6.666666666666735130e-01
+    val Lg2 = 3.999999999940941908e-01
+    val Lg3 = 2.857142874366239149e-01
+    val Lg4 = 2.222219843214978396e-01
+    val Lg5 = 1.818357216161805012e-01
+    val Lg6 = 1.531383769920937332e-01
+    val Lg7 = 1.479819860511658591e-01
+
+    val bits0 = JDouble.doubleToRawLongBits(x)
+    val high0 = (bits0 >>> 32).toInt
+    val low = bits0.toInt
+
+    val (scaledX, high, k0) = if (high0 < 0x00100000) {
+      if (((high0 & 0x7fffffff) | low) == 0) {
+        return -Two54 / 0.0  // log(+-0)=-inf
+      }
+      if (high0 < 0) {
+        return (x - x) / 0.0  // log(-#) = NaN
+      }
+      val scaledX = x * Two54
+      val scaledHigh = (JDouble.doubleToRawLongBits(scaledX) >>> 32).toInt
+      (scaledX, scaledHigh, -54)
+    } else {
+      (x, high0, 0)
+    }
+    if (high >= 0x7ff00000) { // x is inf or NaN
+      return x + x
+    }
+    val highMantissa = high & 0x000fffff // top 20 bits of mantissa
+
+    /* Find k and f such that `x = 2^k * (1+f)`,
+     * where `sqrt(2)/2 < 1+f < sqrt(2)`.
+     *
+     * First, check if mantissa (1+f) is larger than √2 by computing
+     * `val i = (highMantissa + 0x95f64) & 0x100000` where
+     * 0x6a09c is the top 20 bits approximation of the mantissa of √2
+     * and, 0x100000 - 0x6a09c = 0x95f64
+     * If mantissa is larger than √2, i is 0x100000, otherwise it's 0.
+     */
+    val i = (highMantissa + 0x95f64) & 0x100000
+    val bits = JDouble.doubleToRawLongBits(scaledX)
+    /* `(highMantissa | (i ^ 0x3ff00000)).toLong << 32)` sets the exponent to
+     * 0 if the original mantissa (1+f) was smaller than √2, otherwise -1.
+     * newMantissa =
+     *   2^0 * (1+f)    (when (1+f) < √2)
+     *   2^(-1) * (1+f) (when (1+f) >= √2)
+     *
+     * Now, x = 2^newExponent * newMantissa
+     */
+    val newMantissa =
+        JDouble.longBitsToDouble(((highMantissa | (i ^ 0x3ff00000)).toLong << 32) | (bits & 0x00000000ffffffffL))
```

Comment:
```suggestion
    val newMantissa =
      JDouble.longBitsToDouble(((highMantissa | (i ^ 0x3ff00000)).toLong << 32) | (bits & 0xffffffffL))
```
We typically don't write all the leading 0's in hex literals.

Reply:
In fact, consider factoring out
```scala
@inline private def makeDouble(high: Int, low: scala.Long): scala.Double =
  JDouble.longBitsToDouble((high.toLong << 32) | (low & 0xffffffffL))
```
because you're using that several times across the 3 log functions.